File: temp.cs
using System;
using System.Collections.Generic;

namespace FractalTerrain {
    class Point {
        public double x;
        public double y;
    }

    class Program {
        static void Main(string[] args) {
            int iterations = 8;
            double roughness = 0.5;
            double displacementFactor = 1.0;
            Random generator = new Random();
            double[] ys = { 0.0, 0.0 };

            for (int i = 0; i < iterations; i++) {
                ys = Split(ys, displacementFactor, generator);
                displacementFactor *= roughness;
            }

            Point[] points = ComposeCoordinatePairs(ys);
            foreach (Point p in points) {
                Console.WriteLine("{0:0.000} {1:0.000}", p.x, p.y);
            }
        }

        static double[] Split(double[] ys, double displacementFactor, Random generator) {
            double[] newYs = new double[2 * ys.Length - 1];

            for (int i = 0; i < ys.Length - 1; i++) {
                double midpoint = (ys[i] + ys[i + 1]) / 2.0;
                double displacement = generator.NextDouble() * displacementFactor;
                newYs[2 * i] = ys[i];
                newYs[2 * i + 1] = midpoint + displacement;
            }

            return newYs;
        }

        static Point[] ComposeCoordinatePairs(double[] ys) {
            double increment = 1.0 / (ys.Length - 1);
            Point[] points = new Point[ys.Length];

            for (int i = 0; i < ys.Length; i++) {
                points[i] = new Point();
                points[i].x = increment * i;
                points[i].y = ys[i];
            }

            return points;
        }
    }
}

File: temp.py
import numpy as np
from matplotlib import pyplot as plt
from scipy import interpolate

gen = np.random.default_rng(9282003)
    
def split(ys : list[float], displacementFactor : float) -> list[float]:
    newYs = []

    for i in range(len(ys) - 1):
        midpoint = (ys[i] + ys[i + 1]) / 2.0
        displacement = gen.random() * displacementFactor
        newYs.append(ys[i])
        newYs.append(midpoint + displacement)
    
    newYs.append(ys[-1])

    return newYs

def randomSampler(xs, ys):
    cumulative = np.cumsum(ys)
    f = interpolate.interp1d(cumulative/cumulative.max(), xs)
    def sampler():
        return f(gen.random()) 
    return sampler


if __name__ == '__main__':
    iterations = 8
    roughness = .3 # Changes bumpiness
    displacementFactor = 1.0 # Changes max y
    ys = [0.0, 0.0]

    for i in range(iterations):
        ys = split(ys, displacementFactor)
        displacementFactor *= roughness
    xs = np.linspace(0, 1, len(ys))

    generator = randomSampler(xs, ys)
    samples = [generator() for _ in range(10000)]
    fig, axes = plt.subplots(2, 2)
    # fig, axes = plt.subplots(2, 2)
    axes[0,0].hist(samples, bins=100)
    axes[0,1].plot(xs, ys)

    iterations = 8
    roughness = .5 # Changes bumpiness
    displacementFactor = 1.0 # Changes max y
    ys = [0.0, 0.0]

    for i in range(iterations):
        ys = split(ys, displacementFactor)
        displacementFactor *= roughness
    xs = np.linspace(0, 1, len(ys))

    generator = randomSampler(xs, ys)
    samples = [generator() for _ in range(10000)]
    # fig, axes = plt.subplots(1, 2)
    # fig, axes = plt.subplots(2, 2)
    axes[1,0].hist(samples, bins=100)
    axes[1,1].plot(xs, ys)
    plt.show()
